presto-main/src/main/java/com/facebook/presto/operator/aggregation/arrayagg/SetUnionFunction.java

```diff
@@ -0,0 +1,160 @@
+import com.facebook.presto.operator.aggregation.AggregationMetadata.ParameterMetadata;
+import com.facebook.presto.operator.aggregation.AggregationUtils;
+import com.facebook.presto.operator.aggregation.GenericAccumulatorFactoryBinder;
+import com.facebook.presto.operator.aggregation.InternalAggregationFunction;
+import com.facebook.presto.operator.aggregation.SetOfValues;
+import com.facebook.presto.operator.aggregation.state.SetAggregationState;
+import com.facebook.presto.operator.aggregation.state.SetAggregationStateFactory;
+import com.facebook.presto.spi.function.AccumulatorState;
+import com.facebook.presto.spi.function.AccumulatorStateFactory;
+import com.facebook.presto.spi.function.AccumulatorStateSerializer;
+import com.google.common.collect.ImmutableList;
+
+import java.lang.invoke.MethodHandle;
+import java.util.List;
+
+import static com.facebook.presto.common.type.TypeSignature.parseTypeSignature;
+import static com.facebook.presto.operator.aggregation.AggregationMetadata.ParameterMetadata.ParameterType.BLOCK_INDEX;
+import static com.facebook.presto.operator.aggregation.AggregationMetadata.ParameterMetadata.ParameterType.NULLABLE_BLOCK_INPUT_CHANNEL;
+import static com.facebook.presto.operator.aggregation.AggregationMetadata.ParameterMetadata.ParameterType.STATE;
+import static com.facebook.presto.spi.function.Signature.typeVariable;
+import static com.facebook.presto.util.Reflection.methodHandle;
+
+public class SetUnionFunction
+        extends SqlAggregationFunction
+{
+    public static final SetUnionFunction SET_UNION = new SetUnionFunction();
+
+    private static final String NAME = "set_union";
+    private static final MethodHandle INPUT_FUNCTION = methodHandle(SetUnionFunction.class, "input", Type.class, ArrayType.class, SetAggregationState.class, Block.class, int.class);
+    private static final MethodHandle COMBINE_FUNCTION = methodHandle(SetUnionFunction.class, "combine", SetAggregationState.class, SetAggregationState.class);
+    private static final MethodHandle OUTPUT_FUNCTION = methodHandle(SetUnionFunction.class, "output", SetAggregationState.class, BlockBuilder.class);
+
+    public SetUnionFunction()
+    {
+        super(NAME,
+                ImmutableList.of(typeVariable("T")),
+                ImmutableList.of(),
+                parseTypeSignature("array(T)"),
+                ImmutableList.of(parseTypeSignature("array(T)")));
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "Given a column of array type, return an array of all the unique values contained in each of the arrays in the column";
+    }
+
+    @Override
+    public InternalAggregationFunction specialize(BoundVariables boundVariables, int arity, TypeManager typeManager, FunctionManager functionManager)
+    {
+        Type elementType = boundVariables.getTypeVariable("T");
+        ArrayType arrayType = (ArrayType) typeManager.getParameterizedType(StandardTypes.ARRAY, ImmutableList.of(
```

Comment:
If I do that, I would need to add typeManager as a parameter to generateAggregation, right? Or are you suggesting something different?

Reply:
True. It's ok either ways. Currently you could give `generateAggregation` incompatible `elementType` and `arrayType`, which is the thing I don't like. But since it's a private method I don't have strong opinions either ways.

Reply:
I got rid of the generateAggregation method altogether and inlined it into specialize method